src/lib/branchpredictors/GenericPredictor.cc

```diff
@@ -22,76 +22,55 @@ GenericPredictor::GenericPredictor(ryml::ConstNodeRef config)
   // Create branch prediction structures
   btb_ =
       std::vector<std::pair<uint8_t, uint64_t>>(1 << btbBits_, {satCntVal, 0});
-  // Alter globalHistoryLength_ value to better suit required format in update()
-  globalHistoryLength_ = (1 << globalHistoryLength_) - 1;
+
+  // Generate a bitmask that is used to ensure only the relevant number of
+  // bits are stored in the global history. This is two times the
+  // globalHistoryLength_ to allow rolling back of the speculatively updated
+  // global history in the event of a misprediction.
+  globalHistoryMask_ = (1 << (globalHistoryLength_ * 2)) - 1;
```

Comment:
The 1 is by default interpreted as an `int`, so will overflow once `globalHistoryLength_ * 2` >= 32.
In these cases, we should use `1ull`, or an alternative (but messier and I'm guessing slower) is to static_cast<uint64_t>(1).

Also, we need to be careful here as iirc `globalHistoryLength_` has a max of 32, meaning we can be left with the case of `(1 << 64) - 1`. This _should_ work and give a binary value of 64 1s, but is UB as you're overflowing the value to 0 before subtracting one (going out of range).

This will need a further think of how to handle this, but I've not done this enough to have a clean solution at hand. A quick talking point for the meeting on Thursday!

Reply:
Also note I may have missed other cases of this happening - please address any affected code as it's a bit more difficult to see without the full files in front of me.

This only impacts cases where the type we shift is smaller than the type we get out of it, and the overflowing behaviour is when we may shift by the same (or larger) size than the target type.

Reply:
UPDATE: The issue of overflows for unsigned ints is not a problem. After doing more research, over/underflows of unsigned ints is defined behaviour that just leads to the number modulo the type maximum. It's only on _signed_ ints where this causes problems. 

Glad about this as it would have otherwise led to a messier implementation